interfaces/cython/cantera/examples/reactors/ic_engine.py

```
@@ -146,7 +146,7 @@ def piston_speed(t):
             raise 'Error: Refinement limit reached'
```

There is a bug here too, can you fix it by raising a `ct.CanteraError`?

---

And on the line prior, can you change it to use `==` rather than `is`?

---

If that for loop iterates from 0 to 3, will the if condition ever be fulfilled? What is its significance?

---

Indeed, that check makes no sense as written.

---

I'll remove it then.

---

It would be nice if when that loop exited because the limit was reached without success if the script did not continue, and an error was reported to the user. There's also another `is` comparison that should be `==` a few lines below.

---

I'll change `is`  to `==` . And for the error to be reported to the user if limit was reached without success, shouldn't the condition `if n2 == 4` be after the loop and not inside it.

---

The more I think about it, the more I think the whole logic in that loop is borked. What it does right now is try to advance the reactor to `t_i`. If it succeeds, it tries again up to 3 more times (effectively not doing anything other than wasting time). If it fails, the max time step is reduced and the loop goes around to `advance` the simulation again. If it fails 4 times, it just silently continues with the script, even though the simulation isn't complete. Oy.

The most straightforward thing to do is add a Boolean variable before the start of the loop called `solved` initially `False`, set `solved` to `True` and `break` the loop if the `advance` doesn't `raise`, and check the value outside the loop to `raise` an error. The fun way is to use the Python [loop-else](http://python-notes.curiousefficiency.org/en/latest/python_concepts/break_else.html) clause to raise the error if the loop doesn't get broken 😄 (I recommend the straightforward way).

---

The logic is even more borked than that. The real problem is from changing model parameters in the loop without calling `sim.reinitialize()`. If that problem is fixed, there won't be any need for any of this try/except nonsense. But that seems well outside the scope of this PR.

---

@speth Currently i am unfamiliar with the working of  `sim.reinitialize()`. I implemented the changes suggested by @bryanwweber .